drivers: crypto: mbedtls-shim: replace legacy crypto with PSA API

valeriosetti opened the pull request:
This PR replaces usage of Mbed TLS' legacy crypto with PSA API in the shim driver.
Tests are also updated and can be used to verify this PR locally with:

`twister -c -p native_sim -s sample.drivers.crypto.mbedtls -s crypto.aes.mbedtls_shim -s crypto.hash.mbedtls_shim`

valeriosetti committed:
drivers: crypto: mbedtls_shim: replace legacy hash API with PSA API

Replace usage of legacy crypto hash functions from Mbed TLS with PSA API.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: crypto: mbedtls_shim: replace legacy cipher API with PSA API

Replace legacy crypto API for cipher (ECB, CBC) and AEAD (CCM, GCM) with
PSA API.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: crypto: crypto_hash: remove unnecessary Kconfig

Cipher related Kconfigs are unnecessary for this test since it's only
evaluating Hash algorithms.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on drivers/crypto/Kconfig:
```diff
@@ -24,9 +24,10 @@ source "subsys/logging/Kconfig.template.log_config"
 config CRYPTO_MBEDTLS_SHIM
 	bool "MbedTLS shim driver [EXPERIMENTAL]"
 	select MBEDTLS
+	select MBEDTLS_PSA_CRYPTO_C
```
Can now use `PSA_CRYPTO`? Or is this really meant to remain Mbed TLS?

frkv replied:
Discussed offline. `select PSA_CRYPTO` is a reasonable approach to this functionality we expect to be removed as part of the Mbed TLS legacy crypto deprecation

tomi-font commented on drivers/crypto/Kconfig:
```diff
@@ -24,9 +24,10 @@ source "subsys/logging/Kconfig.template.log_config"
 	select MBEDTLS
+	select MBEDTLS_PSA_CRYPTO_C
+	imply PSA_WANT_ALG_SHA_256
+	imply PSA_WANT_KEY_TYPE_AES
```
`select`?

tomi-font replied:
Ah I see, you added the possibility to enable only part of the stuff.

tomi-font commented on samples/drivers/crypto/prj.conf:
```diff
@@ -4,3 +4,5 @@ CONFIG_LOG=y
 
 CONFIG_MAIN_STACK_SIZE=4096
+
+CONFIG_DEBUG=y
```
fail

valeriosetti committed:
drivers: crypto: mbedtls_shim: rename module

Use full name for the module so "mtls" -> "mbedtls"

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: crypto: mbedtls_shim: general renaming

This commit just implements some renaming:
* mbedTLS -> Mbed TLS
* MbedTLS -> Mbed TLS
* MTLS_ -> MBEDTLS_
* mtls_ -> mbedtls_
* mtls_ccm -> ccm_ctx
* mtls_gcm -> gcm_ctx
* mtls_aes -> aes_ctx
* mtls_sha256 -> sha256_ctx
* mtls_sha512 -> sha512_ctx

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: crypto: mbedtls_shim: add _cipher_ to cipher ops

Another small renaming for cipher ops that just adds "_cipher_" in the
function's name.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: crypto: mbedtls_shim: move mbedtls_get_unused_session_index()

Move mbedtls_get_unused_session_index() at the top of the file and leave
cipher and hash functions below.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented on drivers/crypto/crypto_mtls_shim.c:
I can take a look, but IIRC I just renamed the file locally and added the change to git. I don't know why it's shown as deletion+addition

tomi-font replied:
yeah you need to use `git mv`

valeriosetti replied:
Actually doing `git show b87b9842c16501c98c468f853dcbc2681927a4eb` it shows as the file being renamed. I think it's shown as deletion+addition only if you compare this PR against its merge-base point with `main` due to the following changes.

tomi-font replied:
Hmm true, you have a rename commit, oh well...

valeriosetti committed:
mgmt: osdp: remove unnecessary cipher selection in mbedtls-shim driver

Cipher support is already implied by default so there is no need to
select it explicitly.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
net: lib: midi2: remove unnecessary cipher selection in mbedtls-shim driver

Cipher support is already implied by default so there is no need to
select it explicitly.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: net: ieee802154: remove cipher selection in mbedtls-shim driver

Cipher support is already implied by default so there is no need to
select it explicitly.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

r2r0 commented on drivers/crypto/crypto_mbedtls_shim.c:
```diff
@@ -354,114 +284,70 @@ static int mbedtls_cipher_session_setup(const struct device *dev,
-		LOG_ERR("Unhandled mode");
-		ret = -EINVAL;
+		LOG_ERR("Unsupported mode");
+		session->in_use = false;
```
In `mbedtls_cipher_session_free` access to `session->in_use` is performed with mutex lock - why it is not the same here?
Or other way - consider making `in_use` atomic instead and then session release will not need locking.

r2r0 commented on drivers/crypto/crypto_mbedtls_shim.c:
```diff
@@ -590,10 +566,9 @@ static int mbedtls_hash_session_free(const struct device *dev, struct hash_ctx *
 	struct mbedtls_shim_session *mbedtls_session =
```
nit:
Is it different type of session object that in `mbedtls_cipher_session_free` - if not then why not make naming more consistent and use the same name for session object?

valeriosetti committed:
drivers: crypto: mbedtls_shim: improve session handling

Add a new function to free an mbedtls_shim session:
'mbedtls_free_session()'. Together with 'mbedtls_get_unused_session()',
these will be the only entry points to acquire/release sessions.
In this way it's easier to ensure that the mutex is properly held while
setting the 'in_use' parameter.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

ceolin reviewed (approved):
Changes look good. Thanks for all symbol cleanup as well.
